### PR TITLE
metrics: accept any viable target in DvcRepo

### PR DESCRIPTION
**Pull request by pared**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

**Comment by pared** [comment]
@jorgeorpinel Do you think its worth mentioning in docs that from this change on `dvc metrics` will work with any file if its provided via `targets`?

File: dvc/repo/metrics/show.py [hunk 1]
```diff
@@ -22,17 +22,29 @@ def _collect_metrics(repo, targets, recursive):
...
-
-    def _filter(path_info):
-        for info in target_infos:
-            func = path_info.isin_or_eq if recursive else path_info.__eq__
-            if func(info):
-                return True
-        return False
-
-    return list(filter(_filter, metrics))
+    else:
```
Comment (efiop) [inline dvc/repo/metrics/show.py:25]: This `else` is not needed, as you `return` above anyway

File: dvc/repo/metrics/show.py [hunk 2]
```diff
@@ -22,17 +22,29 @@ def _collect_metrics(repo, targets, recursive):
...
+                    )
+                    if func(info):
+                        return True
+                return False
+
+            result.extend(list(filter(_filter, metrics)))
+
+        tree = RepoTree(repo)
+        for target in targets:
+            if tree.exists(target) and not tree.isdir(target):
```
Comment (efiop) [inline dvc/repo/metrics/show.py:45]:
```suggestion
            if tree.isfile(target):
```
?

Comment (efiop) [reply dvc/repo/metrics/show.py:45]: Not strictly equivalent, but wonder if that's what you meant here.

File: tests/func/metrics/test_show.py [hunk 3]
```diff
@@ -110,3 +110,31 @@ def test_missing_cache(tmp_dir, dvc, run_copy_metrics):
     remove(stage.outs[0].cache_path)
 
     assert dvc.metrics.show() == {"": {"metrics.yaml": 1.1}}
+
```
Comment (efiop) [inline tests/func/metrics/test_show.py:113]: Let's also add a test for showing something that is not metric in other branches.

File: tests/func/metrics/test_show.py [hunk 4]
```diff
@@ -110,3 +110,31 @@ def test_missing_cache(tmp_dir, dvc, run_copy_metrics):
...
+    run_copy_metrics("metrics_t.yaml", metric1, metrics=[metric1])
+    run_copy_metrics("metrics_t.yaml", metric2, metrics=[metric2])
+
+    result = dvc.metrics.show(
+        targets=["metrics_t.yaml", "metrics"], recursive=True
+    )
+    assert result == {
+        "": {
+            "metrics/metric1.yaml": {"foo": 1.1},
+            "metrics/metric2.yaml": {"foo": 1.1},
```
Comment (efiop) [inline tests/func/metrics/test_show.py:137]: Will it be / for windows too though?

File: tests/func/metrics/test_show.py [hunk 5]
```diff
@@ -110,3 +110,31 @@ def test_missing_cache(tmp_dir, dvc, run_copy_metrics):
...
+    }
+
+
+def test_show_non_metric_and_recursive_metric(tmp_dir, dvc, run_copy_metrics):
+    tmp_dir.gen({"metrics_t.yaml": "foo: 1.1", "metrics": {}})
+
+    metric1 = os.fspath(tmp_dir / "metrics" / "metric1.yaml")
+    metric2 = os.fspath(tmp_dir / "metrics" / "metric2.yaml")
+    run_copy_metrics("metrics_t.yaml", metric1, metrics=[metric1])
+    run_copy_metrics("metrics_t.yaml", metric2, metrics=[metric2])
```
Comment (efiop) [inline tests/func/metrics/test_show.py:129]: These are not really related to this PR, right? Or did you want to test if recursive works for dirs that don't have official metrics? Just a bit confused.

Comment (pared) [reply tests/func/metrics/test_show.py:129]: I wanted to verify whether `show` will work as intended if we mix `non-metric` and `metric` approach in case of `recursive=true`

File: dvc/repo/metrics/show.py (hunk 2 above)
Comment (pared) [reply dvc/repo/metrics/show.py]: Well I wanted to omit directories in case of `recursive`, so applying some logic `isfile` seems like a better check.

File: tests/func/metrics/test_show.py (hunk 5 above)
Comment (efiop) [reply tests/func/metrics/test_show.py:140]: Makes sense.

File: dvc/repo/metrics/show.py [hunk 6]
```diff
@@ -23,16 +23,26 @@ def _collect_metrics(repo, targets, recursive):
...
 
-    def _filter(path_info):
-        for info in target_infos:
-            func = path_info.isin_or_eq if recursive else path_info.__eq__
-            if func(info):
-                return True
-        return False
+        def _filter(path_info):
+            for info in target_infos:
+                func = path_info.isin_or_eq if recursive else path_info.__eq__
```
Comment (efiop) [inline dvc/repo/metrics/show.py:34]: Does this make any sense now? You are under `if recursive` already

File: dvc/repo/metrics/show.py [hunk 7]
```diff
@@ -23,16 +23,25 @@ def _collect_metrics(repo, targets, recursive):
...
+                    return True
+            return False
 
-    return list(filter(_filter, metrics))
+        result.extend(list(filter(_filter, metrics)))
+
+    tree = RepoTree(repo)
+    for target in targets:
+        if tree.isfile(target):
+            result.append(PathInfo(target))
```
Comment (efiop) [inline dvc/repo/metrics/show.py:43]: Seems like you can simply reuse `target_infos` from above.

File: dvc/repo/metrics/show.py [hunk 8]
```diff
@@ -23,16 +23,23 @@ def _collect_metrics(repo, targets, recursive):
     if not targets:
         return list(metrics)
 
+    result = []
     target_infos = [PathInfo(os.path.abspath(target)) for target in targets]
+    if recursive:
```
Comment (efiop) [inline dvc/repo/metrics/show.py:28]: Will recursive work with unofficial metrics? I don't see a test for it and the code seems like it won't work.

Comment (pared) [reply dvc/repo/metrics/show.py:28]: I am not sure how that would work, I mean, in that case we would need to try to show all files for particular directory, since they are not marked as metrics. So the question here is whether we want to support that.

Comment (efiop) [reply dvc/repo/metrics/show.py:28]: @pared That seems like a natural approach. You have a test with official metrics dir in this PR, what changes if they are not declared as official metrics?

File: dvc/repo/metrics/show.py [hunk 9]
```diff
@@ -23,16 +23,23 @@ def _collect_metrics(repo, targets, recursive):
...
+                if path_info.isin_or_eq(info):
+                    return True
+            return False
 
-    return list(filter(_filter, metrics))
+        result.extend(list(filter(_filter, metrics)))
+
+    tree = RepoTree(repo)
+    for target_info in target_infos:
+        if tree.isfile(target_info):
```
Comment (efiop) [inline dvc/repo/metrics/show.py:40]: What if a target is a file and `recursive==True`? Will it be added to the `result` twice?

File: dvc/repo/metrics/show.py (hunk 8 above)
Comment (pared) [reply dvc/repo/metrics/show.py:28]: On the other hand, it kinda makes sense to have special dir for metrics and check all of them out. Ill try to modify the change.

Comment (efiop) [reply dvc/repo/metrics/show.py:28]: Also note that we try to parse metric files, so if something is not a metric we'll just ignore it and won't show it.

File: dvc/repo/metrics/show.py (hunk 9 above)
Comment (efiop) [reply dvc/repo/metrics/show.py:40]: This will probably become obsoleted once you implement recursive for unofficial metrics above...

File: tests/func/metrics/test_show.py [hunk 10]
```diff
@@ -110,3 +110,42 @@ def test_missing_cache(tmp_dir, dvc, run_copy_metrics):
...
+    tmp_dir.scm_gen("metrics.yaml", "foo: 1.1", commit="init")
+    with tmp_dir.branch("branch", new=True):
+        tmp_dir.scm_gen("metrics.yaml", "foo: 2.2", commit="other")
+
+    assert dvc.metrics.show(targets=["metrics.yaml"], revs=["branch"]) == {
+        "workspace": {"metrics.yaml": {"foo": 1.1}},
+        "branch": {"metrics.yaml": {"foo": 2.2}},
+    }
+
+
```
Comment (efiop) [inline tests/func/metrics/test_show.py:133]: Missing a test for recursive directory with unofficial metrics.

Comment (pared) [reply tests/func/metrics/test_show.py:133]: Since both official and unofficial metrics are gathered the same way now, I would argue that `test_non_metric_and_recurisve_show` already covers that. Anyhow, I modified it so that it checks both official and unofficial metrics.

Comment (efiop) [reply tests/func/metrics/test_show.py:133]: Thank you! :pray:

**Review by efiop — APPROVED** [review]
Thanks!

**Comment by jorgeorpinel** [comment]
> Do you think its worth mentioning in docs

Yes! Very much 🙂 — thanks for submitting a docs PR BTW.
